src/lib/metrics.rs

```diff
@@ -204,6 +204,94 @@ pub struct DemuxedGroupMetrics<'a> {
     pub sample_barcode_hop_tracker: Option<SampleBarcodeHopTracker<'a>>,
 }
 
+/// Write the per sample metrics.
+fn write_per_group_metrics<P: AsRef<Path>>(
+    samples: &[SampleMetadata],
+    total_templates: usize,
+    per_group_metrics: &[DemuxedGroupSampleMetrics],
+    output_dir: &P,
+    filename: &str,
+) -> Result<()> {
+    // Get the group with the highest template count. Don't include unmatched when determining
+    // the "best" barcode.
+    let best_barcode_count = per_group_metrics[0..per_group_metrics.len() - 1]
+        .iter()
+        .max_by_key(|s| s.total_matches)
+        .map_or(0, |s| s.total_matches);
+
+    // Build per group metrics to output.  We need to derive (e.g. ratio, percentage) values here.
+    let per_group_metrics: Vec<SampleMetricsProcessed> = samples
+        .iter()
+        .zip(per_group_metrics.iter())
+        .map(|(sample, metrics)| metrics.as_processed(total_templates, best_barcode_count, sample))
+        .collect();
+
+    // Finally output it to a file
+    let output_path = output_dir.as_ref().join(filename);
+    let delim = DelimFile::default();
+    delim.write_tsv(&output_path, per_group_metrics)?;
+
+    Ok(())
+}
+
+/// Builds a new set of "samples" and demux metrics by grouping input samples by project.  Each
+/// new sample corresponds to a unique project, and the associated metric are the sum of values
+/// for samples with the same project.  The undetermined sample is not grouped with any other
+/// sample.
+fn build_per_project_metrics(
+    samples: &[SampleMetadata],
+    per_sample_metrics: &[DemuxedGroupSampleMetrics],
+) -> Result<(Vec<SampleMetadata>, Vec<DemuxedGroupSampleMetrics>)> {
+    let mut project_to_metric: AHashMap<Option<&String>, DemuxedGroupSampleMetrics> =
+        AHashMap::new();
+    let mut project_to_sample: AHashMap<Option<&String>, SampleMetadata> = AHashMap::new();
```

Comment:
You can just use the stock HashMap here ... AHashMap is useful in performance sensitive contexts, but not here.